### PR TITLE
PX-245: Use a FormData object and XHR request to upload files rather than iframes

### PR DESCRIPTION
**Pull request by rhoggSugarcrm**
Using this instead of #13 because of #21

### CODE REVIEW

File: test/client.js [hunk 1]
```diff
@@ -993,17 +993,11 @@ describe('Api client', function () {
 
             let xhr = this.sandbox.useFakeXMLHttpRequest();
 
-            const fileObject = new File(['I am a file'], 'someFile.txt');
```
Comment (rhoggSugarcrm) [inline test/client.js]: Figured out why this doesn't work: IE and Edge don't support the `File` constructor. https://caniuse.com/#feat=fileapi

File: src/client.js [hunk 2]
```diff
@@ -1287,36 +1281,29 @@ function SugarApi(args) {
          * @param {Object} [$files] jQuery/Zepto DOM elements that carry the files to upload.
          * @param {Object} [callbacks] callback object.
          * @param {Object} [options] Request options hash.
+         * @param {boolean} [options.iframe=true] If `true`, use iframe transport.
```
Comment (rhoggSugarcrm) [inline src/client.js]: OOPS! I don't think this option even has any effect any more. Make sure to check that before sending to CR.

File: src/client.js [hunk 3]
```diff
@@ -652,6 +646,7 @@ function SugarApi(args) {
             };
 
             //Process the iframe transport request
+            // FIXME: should we have jquery.iframe-transport as a dependency/peerDependency?
```
Comment (rhoggSugarcrm) [inline src/client.js]: NOTE TO SELF: DELETE THIS.

File: src/client.js [hunk 4]
```diff
@@ -1287,36 +1281,27 @@ function SugarApi(args) {
...
-                // pass OAuth token as GET-parameter during file upload.
-                // otherwise, in case if file is too large, the whole request body may be
-                // ignored by interpreter together with the token
-                options = options || {};
-                options.passOAuthToken = true;
+            let fd = new FormData();
+
+            if (data.field && $files && !_.isUndefined($files[0]) && $files[0].files && $files[0].files.length > 0) {
+                fd.append(data.field, $files[0].files[0]);
             }
```
Comment (nportik-sugarcrm) [inline src/client.js]: One small thing here: the length check against 0 could be removed (any number greater than 0 should give a `true` and if it is 0 it should be automatically `false`).

Also I wonder if the if statement above could be brought under a more readable format, something like:
`
if (data.field && $files) { // we check if there was anything sent at all
    let attachedFile = $files[0];
    if (!_.isUndefined(attachedFile) && attachedFile.files && attachedFile.files.length) { //we check if we really have files to work with
        fd.append(data.field, attachedFile.files[0]);
    }
}
`

File: src/client.js [hunk 5]
```diff
@@ -935,11 +934,6 @@ function SugarApi(args) {
                 params.delete_if_fails = true;
             }
 
-            // This is for BWC only. Don't document it and remove as soon as 6.7 is decommissioned.
```
Comment (dwheeler) [inline src/client.js]: Not exactly on time, but good to see this comment getting executed on :)

**Pull request merged**